### ISSUE TITLE
PingPong is not functioning as expected

### ISSUE THREAD

**Issue by Camille92** [ctubio/Krypto-trading-bot#694]
Hi Carles and everyone,

So as I'm coming back to more traditional MM strategies, I realized that PingPong is not functioning as it should. (At least in my humble view).

In PingPong, the Pong should be very stable (especially if bw? is deactivated).

**If your ping is a buy:**
Pong_price = Ping_price (AKA last recorded buy) + Pong width

**If your ping is a sell:**
Pong_price = Ping_price (AKA last recorded sell) - Pong width

**This should work for whatever size is your ping and be invalidated only if:**
For selling pong:
If FV < Ping - Width PONG

For buying pong:
If FV > Ping + Width PONG

I realize it might sound a bit confusing with all the width ping and pong so I give you an example:

Configuration: Market BTC/USD, Width Ping = 5, width Pong = 20 and bw? is off:

You make your first buy at 7000 USD, so it's a Ping:
Now the bot will try to sell at 7020 (Ping + width pong) (if that respects the width ping of 5).

Everything goes well, you sell at 7020, you're going to try to buy back at 7000.
However, the price keeps rising and FV = 7030, in this case, you're still trying to buy at 7000.

Only when FV gets over 7040 (FV > Ping price + width pong), that you're now buying at width PING again.
Let's say that you bought at 7040, you're now trying to sell at 7060 and a new cycle of PingPong starts again.

I hope it's clear and we can fix it :)

Let me know if you have any question!

Best,

Camille

PS: Thank you for unlocking the instance in HitBtc

**Comment by Camille92** [comment]
![capture d ecran 2018-03-30 a 22 04 59](https://user-images.githubusercontent.com/6861590/38152139-8b609afe-3466-11e8-9023-ae82870490d7.png)


I put the same settings as described in the bot as described, and in theory, I should be selling at 7095.7!

**Comment by donaldforest** [comment]
Want to support Camille.. I stop get profit from hardcore MM without safety.
I not sure about respecting pong width, but most interested part in this ticket, like for me, is a solution to reset old ping when we lost price in another direction we hope. I mean:

> Only when FV gets over 7040 (FV > Ping price + width pong), that you're now buying at width PING again. Let's say that you bought at 7040, you're now trying to sell at 7060 and a new cycle of PingPong starts again.

**Comment by Camille92** [comment]
And coming back to pingpong, if a trade is made respecting Pong width, then the trade limit should be back to 0!

Donald, if we don't respect pong width, why bothering having it?

**Comment by donaldforest** [comment]
> Donald, if we don't respect pong width, why bothering having it?
I prefer sold with any Agressive "profit" >= 0. But as I told, very important and interested your reset ping rule when price go in another direction.

**Comment by Camille92** [comment]
Fair enough! that's why pong at fair / aggressive exists!

I'm trying to find where PingPong is defined but I don't find or understand it.

Is it in https://github.com/ctubio/Krypto-trading-bot/blob/master/src/server/qe.h ?

**Comment by Camille92** [comment]
I meant not respecting Pong width from FV but pong width from the last buy or sell ;)

**Comment by Camille92** [comment]
As a reminder, it is how the code looked like in Tribeca:

```
        if (params.mode === Models.QuotingMode.PingPong) {
          if (unrounded.askSz && safety.buyPing && unrounded.askPx < safety.buyPing + params.width)
            unrounded.askPx = safety.buyPing + params.width;
          if (unrounded.bidSz && safety.sellPong && unrounded.bidPx > safety.sellPong - params.width)
            unrounded.bidPx = safety.sellPong - params.width;

```

https://github.com/michaelgrosner/tribeca/blob/master/src/service/quoting-engine.ts

**Comment by Camille92** [comment]
https://github.com/michaelgrosner/tribeca/commit/0f725a5340dbbac2e5c99cb49f0b02d41d5ff74b#diff-663fe5842714a63c797cc7807ac7df2c

**Comment by bdbabiak** [comment]
The code does appear in qe.h it appears to me. Your ideas are very good.

**Comment by ctubio** [comment]
@Camille92 we have this calcs at https://github.com/ctubio/Krypto-trading-bot/blob/master/src/server/qe.h#L219-L220 but they are only applied for Boomerang and AK47 not for pingpong (because `if (!qp._matchPings) return;`; let me fix this little mess at #754 

thanks for all the info''¡
